DESIGN.md

```diff
@@ -73,77 +157,225 @@ Flow control support:
 
-- buffering, sampling, throttling, windowing, dropping, etc
-- temporal and count-based strategies
-- It does not support pull-based consumer-driven flow control.
 
 ##### Disposable
 
 A type representing work that can be cancelled or disposed.
 
-### Behavior
+Examples:
 
-##### Creation
+- An `Observable.subscribe` passes a `Disposable` to the `Observable.onSubscribe` to allow the `Observer` to dispose of the subscription. 
+- A `Scheduler` returns a `Disposable` that you use for disposing of the `Scheduler`.
 
-Creation of a stream falls into the following use cases, all of which should be catered to in API design.
+`Subscription` is a similar type used for lifecycle management on the `Flowable` type with interactive flow control.
 
-- async, hot, push (ie. system or user events)
-- async, cold, push (ie. events resulting from remote system via network connection)
-- sync, cold, pull (ie. iterable, file, range)
-- async, cold, pull (ie. RPC/REST network call, cross-thread queue draining)
- 
-Unknown:
+##### Operator
 
-- hot, pull (what is an example of this?)
+An operator follows a specific lifecycle (union of the producer/consumer contract).
 
-Flow control support:
+- It must propagate the `subscribe` event upstream (to the producer).
+- It must obey the RxJava contract (serialize all events, `onError`/`onComplete` are terminal).
+- If it has resources to cleanup it is responsible for watching `onError`, `onComplete`, and `cancel/dispose`, and doing the necessary cleanup.
+- It must propagate the `cancel/dispose` upstream.
+
+In the addition of the previous rules, an operator for `Flowable`:
+
+- It must propagate/negotiate the `request(n)` event.
+
+
+### Creation
+
+Creation of the various types should be exposed through factory methods that provide safe construction.
+
+```java
+Flowable.create(SyncGenerator generator)
+
+Flowable.create(AsyncGenerator generator)
+
+Observable<T>.create(OnSubscribe<Observer<T>> onSubscribe)
+
+Single<T>.create(OnSubscribe<Single.Subscriber<T>> onSubscribe)
+
+Completable<T>.create(OnSubscribe<Completable.Subscriber<T>> onSubscribe)
+```
+
+### Terminal behavior
+
+A producer can terminate a stream by emitting `onComplete` or `onError`. A consumer can terminate a stream by calling `cancel`/`dispose`.
+
+Any resource cleanup of the source or operators must account for any of these three termination events. In other words, if an operator needs cleanup, then it should register the cleanup callback with `cancel`/`dispose`, `onError` and `onComplete`.
+
+The final `subscribe` will *not* invoke `cancel`/`dispose` after receiving an `onComplete` or `onError`.
+
+### JVM target and source compatibility
+
+The 2.x version will target JDK6+ but the source will be using some of the Java 8 features.
```

Comment:
Okay, I can understand the argument of _its not part of the design_ however there should be a discussion about using lambdas in RxJava source code. I have opened an issue to track this #3805 discussion. For now I think that this line in the design document should explicitly state something to the following effect.

```
* RxJava will be published to `io.reactivex.rxjava:2.x` with bytecode compatible with a Java 1.6 runtime. In order to do so, all references to java library classes and interfaces must be likewise compatible (i.e. no references to Java8 runtime libraries). 
* RxJava source code will be implemented with lambda expressions. 
```

The reason I vote for keeping the bit about lambdas in the design is because it expresses the development environment and constraints that we intend to have on the committers/contributors.